src/checker.rs

```diff
@@ -0,0 +1,88 @@
+use std::error::Error;
+use std::time::Duration;
+
+use reqwest::{ClientBuilder, StatusCode};
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+pub enum CheckResult {
+    SUCCESS,
+    FAILURE(FailureReason),
+    MISSED,
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+pub enum FailureReason {
+    Timeout,
+    DnsError(String),
+    ConfigError(String),
+    Error(String),
+}
+
+pub async fn check_domain(domain: String) -> CheckResult {
+    // TODO: Timeout should come from config
+    let timeout = Duration::new(5, 0);
+    let client_result = ClientBuilder::new().timeout(timeout).build();
+    let client = match client_result {
+        Err(e) => {
+            return CheckResult::FAILURE(FailureReason::ConfigError(e.to_string()));
+        }
+        Ok(client) => client,
+    };
+    let mut response_result = client.head(&domain).send().await;
+    if let Ok(response) = &response_result {
+        if response.status() == StatusCode::METHOD_NOT_ALLOWED {
+            // If this domain doesn't accept HEAD, attempt GET request
+            response_result = client.get(&domain).send().await;
+        }
+    }
+
+    match response_result {
+        Ok(_) => {
+            return CheckResult::SUCCESS;
+        }
+        Err(e) => {
+            if e.is_timeout() {
+                return CheckResult::FAILURE(FailureReason::Timeout);
+            } else {
+                // TODO: More reasons
+
+                let mut source = e.source();
+                while let Some(inner) = source {
+                    if inner.source().is_none() {
+                        break;
+                    }
+                    source = inner.source();
```

Comment:
Rust noob question: wouldn't more idiomatic way be to not use `mut`, but instead shadow it with `let source = inner.source()` in line 54?

Reply:
If you use shadowing the loop doesn't work. You need to mutate the outer `source` variable that loop condition can see.

Reply:
Yeah, this is nicer, and we don't need `source` outside the loop anyway. I think I had that in place from some earlier experiments

Reply:
The entire loop can be improved like so:
```rust
                let mut inner = &e as &dyn Error;
                while let Some(source) = inner.source() {
                    inner = source;

                    // TODO: Would be better to get specific errors without string matching like this
                    // Not sure if there's a better way
                    let inner_message = inner.to_string();
                    if inner_message.contains("dns error") {
                        return CheckResult::FAILURE(FailureReason::DnsError(inner_message));
                    }
                }
```
I'll see if I can find anything about the string matching, that is indeed very unappealing.

Reply:
Yeah, I'm planning on attempting to improve this as well, but so far it looks like reqwest doesn't provide a lot of info here. We can do a lot better with status codes at least, but for other network related issues it might be tricky